core/src/context.hpp

```diff
@@ -54,3 +54,196 @@ json & operator << (json &, km_core_context_item const &);
+ *         * `KM_CORE_STATUS_OK`: On success.
+ *         * `KM_CORE_STATUS_INVALID_ARGUMENT`: If non-optional parameters are
+ *           null.
+ *         * `KM_CORE_STATUS_NO_MEM`: In the event not enough memory can be
+ *           allocated for the output buffer.
+ *
+ * @param context_items A pointer to the start of an array
+ *                      `km_core_context_item`.
+ *
+ * @param out           A pointer to the result variable: A pointer to the start
+ *                      of the `km_core_context_item` array containing a copy of
+ *                      the context. Terminated with a type of `KM_CORE_CT_END`.
+ *                      Must be disposed of with
+ *                      `km_core_context_items_dispose`.
+ */
+km_core_status
+context_get(km_core_context const *context_items,
+                   km_core_context_item **out);
+
+/**
+ * Return the number of items in the context.
+ *
+ * @return size_t The number of items in the context, and will return 0 if
+ *                passed a null `context` pointer.
+ * @param context A pointer to an opaque context object
+*/
+size_t
+context_length(km_core_context *);
+
+/**
+ * Add more items to the end (insertion point) of the context. If these exceed
+ * the maximum context length the same number of items will be dropped from the
+ * beginning of the context.
+ *
+ * @return km_core_status
+ *         * `KM_CORE_STATUS_OK`: On success.
+ *         * `KM_CORE_STATUS_INVALID_ARGUMENT`: If non-optional parameters are
+ *           null.
+ *         * `KM_CORE_STATUS_NO_MEM`: In the event not enough memory can be
+ *           allocated to grow the context buffer internally.
+ *
+ * @param context       A pointer to an opaque context object.
+ * @param context_items A pointer to the start of the `KM_CORE_CT_END`
+ *                      terminated array of `km_core_context_item` to append.
+ */
+km_core_status
+context_append(km_core_context *context,
+                      km_core_context_item const *context_items);
+
+/**
+ * Remove a specified number of items from the end of the context, optionally
+ * add up to the same number of the supplied items to the front of the context.
+ *
+ * @return km_core_status
+ *         * `KM_CORE_STATUS_OK`: On success.
+ *         * `KM_CORE_STATUS_INVALID_ARGUMENT`: If non-optional parameters are
+ *           null.
+ *         * `KM_CORE_STATUS_NO_MEM`: in the event it cannot allocated enough
+ *           memory to grow the context internally.
+ *
+ * @param context       A pointer to an opaque context object.
+ * @param num           The number of items to remove from the end of context.
+ * @param context_items Pointer to the start of the `KM_CORE_CT_END` terminated
+ *                      array of `km_core_context_item` to add to the front. Up
+ *                      to `num` items will be prepended. This may be null if
+ *                      not required.
+ */
+KMN_API
+km_core_status
+context_shrink(km_core_context *context,
```

Comment:
I have some problems with this method. AFAICT it's only used in test code and for that it is ok, but it seems it should be in the test code then. For use as a general method I think the name doesn't match what it does (it also prepends besides removing from the end),  and it's not possible to prepend additional characters so it's less useful for general use.

I think it would be better to split it in two methods, `context_shrink` (with an additional parameter to be able to remove from beginning) and `context_prepend`. In fact, I can use such methods for #10100, so I'll base my changes on this PR and implement the split.

Reply:
OK. I only left it because it was tangled into the unit tests -- but happy to let you iterate improvement on it too.